### PR TITLE
Add support for composite kwargs

### PR DESCRIPTION
**Pull request by hheydary**
Add support for composite keyword arguments.

BUG=b/353293690

### CODE REVIEW

File: ai_edge_torch/convert/test/test_convert.py [hunk 1]
```diff
@@ -306,6 +308,36 @@ def forward(self, x, y):
...
+
+    torch.export.register_dataclass(TestContainer, serialized_type_name="TestContainer")
+
+    class SampleModel(torch.nn.Module):
+
+      def forward(self, x: torch.Tensor, y: torch.Tensor, z: TestContainer):
+        return x + y + z.data_1 + z.data_2[0] + z.data_2[1]
+
+    args_gen = lambda: (torch.randn(10, 10),)
+    kwargs_gen = lambda: dict(
```
Comment (majiddadashi) [inline ai_edge_torch/convert/test/test_convert.py:329]: why not just create the object as opposed to creating an args generator function?

File: ai_edge_torch/convert/test/test_convert.py [hunk 2]
```diff
@@ -306,6 +308,36 @@ def forward(self, x, y):
...
+        y=torch.randn(10, 10),
+        z=TestContainer(
+            data_1=torch.randn(10, 10),
+            data_2=(torch.randn(10, 10), torch.randn(10, 10)),
+        ),
+    )
+
+    model = SampleModel().eval()
+    edge_model = ai_edge_torch.convert(model, args_gen(), kwargs_gen())
+    self.assertIsNotNone(edge_model)
```
Comment (majiddadashi) [inline ai_edge_torch/convert/test/test_convert.py:339]: Please extend this to run the model and make sure the result is correct.

File: ai_edge_torch/convert/test/test_convert.py [hunk 3]
```diff
@@ -306,6 +308,36 @@ def forward(self, x, y):
...
+    """
+    Test converting a simple model with both sample_args and sample_kwargs.
+    """
+
+    @dataclass
+    class TestContainer:
+      data_1: torch.Tensor
+      data_2: Tuple[torch.Tensor, torch.Tensor]
+
+    torch.export.register_dataclass(TestContainer, serialized_type_name="TestContainer")
```
Comment (majiddadashi) [inline ai_edge_torch/convert/test/test_convert.py:321]: in the current state, wondering if we even need this line given that we flatten the args before exporting?

File: ai_edge_torch/convert/conversion_utils.py [hunk 4]
```diff
@@ -79,28 +79,30 @@ def flat_arg_names(self) -> list[str]:
     for i in range(args_spec.num_leaves):
       names.append(f"args_{i}")
 
-    dict_context = (
-        kwargs_spec.context
```
Comment (chunnienc) [inline ai_edge_torch/convert/conversion_utils.py:83]: please respect the if-else branch and special case for defaultdict in your new impl, thanks

File: ai_edge_torch/convert/conversion_utils.py [hunk 5]
```diff
@@ -79,28 +79,30 @@ def flat_arg_names(self) -> list[str]:
...
+        flat_names.append(prefix)
+
+    return flat_names
 
   @property
-  def flat_args(self) -> tuple[torch.Tensor]:
-    return tuple(pytree.tree_flatten(self._normalized_sample_args_kwargs)[0])
+  def flat_args(self) -> tuple[Any]:
+    args, kwargs = self._normalized_sample_args_kwargs
+    return tuple([*args, *kwargs.values()])
```
Comment (chunnienc) [inline ai_edge_torch/convert/conversion_utils.py:105]: Please keep the original impl which returns the flat tensor arg list. The intent is to make `len(flat_args) == len(flat_arg_names)` as a way to debug when there is an error, and implicitly disable internal flattening in torch.export.export as that may have a conflict to this external flattening.

File: ai_edge_torch/convert/test/test_convert.py [hunk 6]
```diff
@@ -306,6 +308,36 @@ def forward(self, x, y):
         model_coverage.compare_tflite_torch(edge_model, model, args_gen, kwargs_gen)
     )
 
+  def test_convert_model_with_args_nested_kwargs(self):
+    """
+    Test converting a simple model with both sample_args and sample_kwargs.
```
Comment (chunnienc) [inline ai_edge_torch/convert/test/test_convert.py:313]: update test docstring

**Review by chunnienc — CHANGES_REQUESTED** [review]

File: ai_edge_torch/convert/test/test_convert.py (hunk 2 above)
Comment (hheydary) [reply ai_edge_torch/convert/test/test_convert.py]: Done

File: ai_edge_torch/convert/conversion_utils.py (hunk 4 above)
Comment (hheydary) [reply ai_edge_torch/convert/conversion_utils.py:83]: I am flattening this context in the updated implementation that fixes the bugs. PTAL.

File: ai_edge_torch/convert/test/test_convert.py (hunk 6 above)
Comment (hheydary) [reply ai_edge_torch/convert/test/test_convert.py]: Done

File: ai_edge_torch/convert/conversion_utils.py (hunk 5 above)
Comment (hheydary) [reply ai_edge_torch/convert/conversion_utils.py:124]: The problem is that during conversion, we use sig.flat_args.
```
  exported_programs: torch.export.ExportedProgram = [
      torch.export.export(sig.module, sig.flat_args, dynamic_shapes=sig.dynamic_shapes)
      for sig in signatures
  ]
  ```
However, with nested arguments, the existing logic would give you the wrong number of keyword arguments.

Please note that on authoring side (i.e, pytorch), the structures are maintained and not flattened.

Comment (chunnienc) [reply ai_edge_torch/convert/conversion_utils.py:124]: > the existing logic would give you the wrong number of keyword arguments.

Could you explain more on this?

> Please note that on authoring side (i.e, pytorch), the structures are maintained and not flattened.

While the flattening would always happen before the conversion. The intent is to make logics here the only source of truth for the flattening rule. The worst case is we deep flatten the kwargs here to get the flat name list and pass the dict values as is to  torch.export.export, but then torch export uses another rule/order to flatten the dict.

Comment (hheydary) [reply ai_edge_torch/convert/conversion_utils.py:124]: Let's take this example:
```
    class SampleModel1(torch.nn.Module):

      def forward(self, x: torch.Tensor, y: torch.Tensor, z: TestContainer):
        return x + y + z.data_1 + z.data_2[0] + z.data_2[1]

    args = (torch.randn(10, 10),)
    kwargs = dict(
        y=torch.randn(10, 10),
        z=TestContainer(
            data_1=torch.randn(10, 10),
            data_2=(torch.randn(10, 10), torch.randn(10, 10)),
        ),
    )
```
On pytorch side, the model would expect **three** arguments (x, y, z). However, the following logic:
```
return tuple(pytree.tree_flatten(self._normalized_sample_args_kwargs)[0])
```
would return **five** values (the flattened args). This triggers an error in `torch.export.export`.

Comment (chunnienc) [reply ai_edge_torch/convert/conversion_utils.py:124]: oh I see, yes you're right on this. Thanks for the explanation.

We may still run into issues with this impl when the model.forward expects kw-only args. But I guess there are no better way to solve this right now.

File: ai_edge_torch/convert/test/test_convert.py [hunk 7]
```diff
@@ -306,6 +310,99 @@ def forward(self, x, y):
...
+    flat_inputs = {
+        "args_0": args[0].numpy(),
+        "y": kwargs["y"].numpy(),
+        "z_data_1": kwargs["z"].data_1.numpy(),
+        "z_data_2_0": kwargs["z"].data_2[0].numpy(),
+        "z_data_2_1": kwargs["z"].data_2[1].numpy(),
+    }
+    self._args_kwargs_test_helper(SampleModel1(), args, kwargs, flat_inputs)
+
+    #### case 2
```
Comment (chunnienc) [inline ai_edge_torch/convert/test/test_convert.py:348]: pls separate this into multiple tests. Name them `test_convert_model_with_args_nested_kwargs_{n}` should be good enough.

File: ai_edge_torch/convert/test/test_convert.py [hunk 8]
```diff
@@ -306,6 +310,99 @@ def forward(self, x, y):
...
+    flat_inputs = {
+        "args_0": args[0].numpy(),
+        "y": kwargs["y"].numpy(),
+        "z_data_1": kwargs["z"].data_1.numpy(),
+        "z_data_2_0_foo": kwargs["z"].data_2[0]["foo"].numpy(),
+        "z_data_2_1": kwargs["z"].data_2[1].numpy(),
+    }
+    self._args_kwargs_test_helper(SampleModel3(), args, kwargs, flat_inputs)
+
+  def _args_kwargs_test_helper(self, model, args, kwargs, flat_inputs):
```
Comment (chunnienc) [inline ai_edge_torch/convert/test/test_convert.py:394]: more meaningful name like `compare_tflite_torch_args_kwargs`

**Review by chunnienc — APPROVED** [review]

File: ai_edge_torch/convert/test/test_convert.py (hunk 2 above)
Comment (majiddadashi) [reply ai_edge_torch/convert/test/test_convert.py]: I think this was marked as resolved by mistake (?)

**Review by majiddadashi — CHANGES_REQUESTED** [review]
Please extend this to allow running the model with the arguments the model can be traced with.

File: ai_edge_torch/convert/test/test_convert.py (hunk 7 above)
Comment (hheydary) [reply ai_edge_torch/convert/test/test_convert.py]: Done.

**Review by majiddadashi — APPROVED** [review]
Approving this PR and @hheydary will send a subsequent PR to modify model.py to make execution consistent.

**Pull request merged**